### ISSUE TITLE
Suggestion For DateTime Offset

### ISSUE THREAD

**Issue by Thraka** [dotnet/runtime#34145]
_From @Emeka-NMC on Thursday, March 26, 2020 5:02:27 AM_

[Suggestion To Add Unique Format-Specifiers For DateTime Offset]

DateTime-FormatString doesn't have unique characters that represents the _offset-hour_ and _offset-minutes_. Instead it uses the following format/convention:
 
| # | Offset Format specifier | Offset Format |
|:---|:---|:---|
| 1 | z | +-H  |
| 2 | zz | +-HH  |
| 3 | zzz | +-HH:mm  |
 
Can we create unique specifiers for offset-hour and another for offset-minute?
For example:
| # | Offset Format specifier | Offset Format |
|:---|:---|:---|
| 1 | r<sup>1</sup> | H  |
| 2 | rr | HH  |
| 4 | n<sup>1</sup> | +/- |
| 4 | e<sup>1</sup> | m |
| 5 | ee | mm |
 
_<sup>1</sup> I chose the last characters in Hou(r), Sig(n), and Minut(e) to represent the offset specifiers._

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: dda93f14-f1b5-fde6-e441-302cdac261ab
* Version Independent ID: 96abc7c4-3719-82da-e271-881d62f0640f
* Content: [Custom date and time format strings](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings#feedback)
* Content Source: [docs/standard/base-types/custom-date-and-time-format-strings.md](https://github.com/dotnet/docs/blob/master/docs/standard/base-types/custom-date-and-time-format-strings.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @Thraka
* Microsoft Alias: **adegeo**

_Copied from original issue: dotnet/docs#17603_

**Comment by Thraka** [comment]
_From @Thraka on Thursday, March 26, 2020 6:13:42 PM_

Sorry, we're just documentation for how it currently works. You should make your suggestion at https://github.com/dotnet/runtime/issues. I'll move this issue there.

**Comment by Clockwork-Muse** [comment]
@Emeka-NMC - Is there a specific problem you're trying to solve by getting split specifiers?

Note that, contrary to what you may expect, it's the _entire_ offset which is the field/unit; specifying anything other than `'zzz'` is almost always going to be wrong due to the existence of multiple timezones/offsets with non-hour amounts (and that half of them are negative, requiring the addition of a sign).

**Comment by Emeka-NMC** [comment]
> _Is there a specific problem you're trying to solve by getting split specifiers?_
@Clockwork-Muse, Yes, date-time-offset can be used to generate filenames/etc, or log. So being able to apply custom format to each field will be useful.
* It supports `-07:00`
* But what if I need `-0700`?

With my original recommendation, this will work: `DateTimeOffset.UtcNow.Offset.ToString("nrree")`
NOTE: I added `r` and `e` just for completeness sake. 

> _I's the entire offset which is the field/unit_
Yes, similar to the HH:mm being the time-component of a date-time-offset struct. Within this time-component, I think of HH as the hour field, and mm as the minute field. These fields are combined and formatted to represent the time-component of date-time-offset. Same applies to the offset-component; it has an hour-field, a minute-field, and a reference-sign-field. They, together gets formatted. Having the ability to format this offset-component (in the same DateTimeOffset.ToString(....) command will make things easier.

**Comment by Clockwork-Muse** [comment]
.... Ignoring the potential in/advisability about disambiguating logs by offset....

> - It supports -07:00
> - But what if I need -0700?

That's a slightly different ask than separate format specifiers for each element, and one I personally would be more amenable to.

> Yes, similar to the HH:mm being the time-component of a date-time-offset struct. Within this time-component, I think of HH as the hour field, and mm as the minute field. 

Right - what I'm getting at is that the offset _is one field_, equivalent to hours or minutes.  Subdividing it - or leaving off the sign - is leaving out part of the field, and results in nonsense.  (This is somewhat in contrast to dates/times that are often _truncated_ -  where smaller fields are omitted; you can't reasonably truncate an offset).

Personally, I feel that the current hour-only specifiers are in error - the offset should never be formatted without minutes, if it has any.

**Comment by Emeka-NMC** [comment]
> _Subdividing it - or leaving off the sign - is leaving out part of the field, and results in nonsense._

Thank you for the response. Doesn't the above statement assumes that the user always intends on retaining the original meaning of date-time-offset? What if the user just needs to use/extract parts of the date-time-offset in order to generate some other data?

I think this [Wiki UTC_Offset](https://en.wikipedia.org/wiki/UTC_offset) article may offer additional context:
> The UTC offset is the difference in hours and minutes from Coordinated Universal Time (UTC) for a particular place and date. It is generally shown in the format ±[hh]:[mm], ±[hh][mm], or ±[hh]. So if the time being described is one hour ahead of UTC (such as the time in Berlin during the winter), the UTC offset would be "+01:00", "+0100", or simply "+01".

i.e. having the ability to represent the _offset field_ portion of date-time-offset _"in the format ±[hh]:[mm], ±[hh][mm], or ±[hh]"_ will be useful.

**Comment by Clockwork-Muse** [comment]
> What if the user just needs to use/extract parts of the date-time-offset in order to generate some other data?

Then you should probably be working with the `Offset` property directly, not monkeying with the output format string.

> i.e. having the ability to represent the offset field portion of date-time-offset "in the format ±[hh]:[mm], ±[hh][mm], or ±[hh]" will be useful.

Sure, but what you're asking for isn't the best way to achieve that.  In particular, there's no reason for the sign to be optional, and minutes can't be left off if there are any.  Oh, and note that leading zeroes aren't removed in their example.

Instead, the format specifiers for the offset should likely be (replacing all instances of the `'z'` specifiers):

| Specifier | Format        | Examples                      |
|-----------|---------------|-------------------------------|
| XX        | ±[hh]:[mm]    | `-700->"-07:00"` `630->"+06:30"` |
| X         | ±[hh]\(:[mm])? | `-700->"-07"` `630->"+06:30"`    |
| xx        | ±[hh][mm]     | `-700->"-0700"` `630->"+0630"`   |
| x         | ±[hh]\([mm])? | `-700->"-07"` `630->"+0630"`     |

**Comment by Emeka-NMC** [comment]
Your proposal makes sense, but doesn't address my original problem/request: _support for granular-format-specifiers for offset field._

**Why?**
I think that your proposal is based on the premise that we should always treat offset as an opaque/single-field object, even though it's actually a `TimeSpan` object (similar to `TimeOfDay`.)


**Format vs Create**
> In particular, there's no reason for the sign to be optional, and minutes can't be left off if there are any.

Doesn't the above statement assume that the resulting (custom-formatted) data still represents a DateTimeOffset? The intent of the user might be to _create_ a new/different data-type, i.e. with parts from a `DateTimeOffset`.

**Summary**
* We can do this: `DateTimeOffset.ToString("'['yyyy','MM','dd','HH','mm','ss','fffffff','zzz']'")`
* And I want to do this (for e.g.): `DateTimeOffset.ToString("'['yyyy','MM','dd','HH','mm','ss','fffffff','n','rr','ee']'")`
  i.e. without the additional step of concatenating the `.Offset.ToString(...)`

**Comment by Clockwork-Muse** [comment]
>I think that your proposal is based on the premise that we should always treat offset as an opaque/single-field object, even though it's actually a `TimeSpan` object (similar to `TimeOfDay`.)

...yes, I believe it should be.  Because the C# object isn't actually important here, but the domain type, which is a single date/time field named "offset".

>Doesn't the above statement assume that the resulting (custom-formatted) data still represents a DateTimeOffset? The intent of the user might be to create a new/different data-type, i.e. with parts from a `DateTimeOffset`.

Depending on what you're doing, again, I'd recommend you convert to the new C# type directly, rather than going through the trouble of serializing it.  You'd only need to turn it into a string if your actual domain type required it.  

And your last example is something I'd recommend against, and am trying to prevent.  Especially separating out the sign like that.

**Comment by joperezr** [comment]
FYI: @tarekgh
